core/src/test/java/edu/wpi/grip/core/http/GripServerTest.java

```diff
@@ -0,0 +1,173 @@
+                server = HttpServer.create(new InetSocketAddress("localhost", GRIP_SERVER_TEST_PORT + portOffset), 1);
+                port = GRIP_SERVER_TEST_PORT + portOffset;
+                break;
+            } catch (IOException ex) {
+                // That port is taken -- keep trying different ports
+            }
+        }
+        MockProjectSettings mockSettings = new MockProjectSettings();
+        mockSettings.setServerPort(port);
+        instance = new GripServer((ignore) -> server, new MockPipeline(mockSettings));
+        instance.start();
+
+        client = new DefaultHttpClient();
+        client.getParams().setParameter(CoreProtocolPNames.PROTOCOL_VERSION, HttpVersion.HTTP_1_1);
+    }
+
+    /**
+     * Test of GetHandler methods, of class GripServer.
+     */
+    @Test
+    public void testGetHandlers() {
+        String path = "/testGetHandlers";
+        GetHandler handler = params -> path;
+        instance.addGetHandler(path, handler);
+        instance.addDataSupplier(path, () -> path);
+        try {
+            String data = doGet(path);
+            assertEquals(data, path);
+        } catch (IOException ex) {
+            fail(ex.getMessage());
+        } finally {
+            instance.removeGetHandler(path); // cleanup
+        }
+    }
+
+    /**
+     * Test of addPostHandler method, of class GripServer.
+     */
+    @Test
+    public void testAddPostHandler() {
+        String path = "/testAddPostHandler";
+        byte[] testBytes = "testAddPostHandler".getBytes();
+        PostHandler handler = bytes -> Arrays.equals(bytes, testBytes);
+        instance.addPostHandler(path, handler);
+        try {
+            doPost(path, testBytes);
+        } catch (IOException e) {
+            fail(e.getMessage());
+        } finally {
+            instance.removePostHandler(handler); // cleanup
+        }
+    }
+
+    /**
+     * Test of data supplier methods
+     */
+    @Test
+    public void testDataSuppliers() {
+        String name = "testDataSuppliers";
+        Supplier<?> supplier = () -> name;
+        instance.addDataSupplier(name, supplier);
+        assertTrue(instance.hasDataSupplier(name));
+        instance.removeDataSupplier(name);
+        assertFalse(instance.hasDataSupplier(name));
+    }
+
+    /**
+     * Test of start and stop methods, of class GripServer.
+     */
+    @Test
+    public void testStartStop() {
+        try {
+            instance.start(); // should do nothing since the server's already running
+            instance.stop();  // stop the server so we know we can start it
+            instance.stop();  // second call should do nothing
+            instance.start(); // restart the server
+            instance.start(); // second call should do nothing
+            instance.restart(); // should stop and then start again
+        } catch (Exception e) {
+            // Starting or stopping when in an invalid state will throw an exception
+            fail(e.getMessage());
```

Comment:
Don't do this. Let Junit catch the exception.
If you do this you loose the stack trace.